### ISSUE TITLE
llm assets

### ISSUE THREAD

**Issue by eagle705** [eagle705/nlp-book#3]
- trends

![image](https://user-images.githubusercontent.com/7252598/231923020-9b1dea6d-8504-410e-a733-ef550b6c2591.png)

- ia3 performance
![image](https://user-images.githubusercontent.com/7252598/232739247-1ee59fc6-7141-49ca-9500-655d03029580.png)

- ICL (GPT-3)
![image](https://user-images.githubusercontent.com/7252598/232743915-d6b309c5-6029-4a30-b9f2-3062206cac44.png)

- llama hf hub
![image](https://user-images.githubusercontent.com/7252598/235445262-0864f7f5-cbbb-4bbb-bc1d-099de5a38879.png)
<img width="314" alt="image" src="https://user-images.githubusercontent.com/7252598/235452521-c5ba4e2c-83ce-40b8-8086-52e9216ff760.png">

- alpaca-lora
![image](https://user-images.githubusercontent.com/7252598/235447484-2f8407a8-3d6e-4fa0-9b48-733566b42c68.png)

![image](https://user-images.githubusercontent.com/7252598/235447454-0d6d33cf-ff17-4b50-9d1a-22db2f6fe845.png)

- bitandbytes
  - https://github.com/TimDettmers/bitsandbytes
![image](https://user-images.githubusercontent.com/7252598/235461109-5861e275-b622-4057-8a8a-1733c93e0bbd.png)

- chinchilla
<img width="480" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/aaaf5425-b0b8-4e8d-9463-54bc65946852">

- redpajama
<img width="785" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/f3422adb-6991-47e1-b7bc-fcdc4f5e270c">

- pre-LN
<img width="311" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/017bed00-0618-47b4-a8d5-70e70666715f">

- Rotary
![image](https://github.com/eagle705/nlp-book/assets/7252598/3f2299c6-e9e0-4620-a322-118ebef90c58)

- LLaMA 성능
<img width="685" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/f062f9e5-119b-4c62-b2f3-0e30afd7666a">

- GPU rank
<img width="1154" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/82ea0e0c-b49c-48dc-ad10-b8c1ee06baf8">

- PIO, DMA
<img width="572" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/bad85369-cdc2-4025-93f0-62b30d713189">
- RDMA
<img width="807" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/e34aa3cc-a98c-45a8-a568-db481cc589ae">
- PCI vs NVLink
<img width="932" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/717f091c-fa52-4771-af5c-e363ca4c84fb">

- 용어
![image](https://github.com/eagle705/nlp-book/assets/7252598/34d654d4-1cc8-44a2-b7eb-322fe55e3899)
![image](https://github.com/eagle705/nlp-book/assets/7252598/4d8bad15-d294-48fb-87d3-9d8c721c88f4)

- 통신 도표 예시
<img width="1026" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/fcf25fba-810f-4720-9283-5cf47820407d">

- MPI
<img width="1007" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/6dcb6614-9ed8-4ea0-9f7b-1abe39224234">
- Message Passing vs Shared Memory
<img width="845" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/517651e4-6a6b-4b97-89f4-e8dec4ecc1c8">

- p2p
![image](https://github.com/eagle705/nlp-book/assets/7252598/640e12dd-0368-46ba-9243-82e5df023849)

- collective communication
<img width="952" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/630e6f86-211b-48ad-b4e6-f02008fe5376">

- Reduce-scatter
<img width="1430" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/bab97630-00d5-433a-8fc5-0f08bcc87884">

## 3D parallel
- DP
<img width="1192" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/420b90ff-04f5-4e7c-abc3-3ecac4ddb5a0">
- forward
<img width="1603" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/c9ec9445-b9bc-420a-ab87-a7925dddd7eb">
- backward
<img width="1195" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/485864b5-e5dd-42d1-b63d-e4975c142d46">

<img width="1125" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/62f1dde8-8cff-4354-bc21-8c7baba56f90">

![image](https://github.com/eagle705/nlp-book/assets/7252598/c2e0ca56-4b44-451d-906c-5ac214c98313)

<img width="461" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/2cc2ef6d-dc45-4f62-8973-04d8bef55cc3">

<img width="1033" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/044eb4d3-bffb-44a3-abd8-330a79507948">

<img width="839" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/d7da4cfa-540b-4971-adf2-dd156616274a">

- 출처: PyTorch & distributed training Tutorial by Shreya Shankar and Karan Goel (https://stanford-cs329s.github.io/syllabus.html)
![image](https://github.com/eagle705/nlp-book/assets/7252598/9ce68cf4-4058-41e5-a8e8-ce8d6368127b)

- DDP
<img width="1299" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/cfc96285-d497-4377-bb45-417776ff5cc1">

![ring_allreduce](https://github.com/eagle705/nlp-book/assets/7252598/38ad9c47-4afc-45d4-93a3-0e75ee131997)

![image](https://github.com/eagle705/nlp-book/assets/7252598/a8db7b97-7ec6-43bc-8488-128989858c0c)

- MP
<img width="1581" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/0b4a1257-aa1b-4b9e-ad24-d2e70a53a061">

<img width="1550" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/85b0404c-38e1-439e-aebd-29671d07031c">

- pp
<img width="999" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/7ab47436-b191-4046-899b-e0df9bc870bb">

<img width="1265" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/07a4a4cc-d1ba-4d60-92ef-4ef59c0ebae8">

<img width="1248" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/56de34ff-5434-486c-9a93-4d7fe826e2bc">

- Activation Memory
<img width="1417" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/08c35441-a6a6-41b0-821e-44a0a5dff920">

<img width="638" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/e264fdb9-59cb-433a-a8e2-1cdb5fbe58c3">

- TP
column parallelism
<img width="1305" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/49e9086b-cb2f-4904-9ecb-6765c0bbc3b4">

row parallelism
<img width="1315" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/a6c7f75f-339d-4d54-87b2-99fa00d92ad0">

MLP

<img width="1555" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/a2cfbff9-d8fb-439f-a882-553171d00fe8">

![image](https://github.com/eagle705/nlp-book/assets/7252598/919a011f-d320-4818-bcf8-9bb1a1d168ec)

Embedding

<img width="1364" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/392fce65-f539-4e25-be10-3b373182117f">

3D parallel
<img width="944" alt="image" src="https://github.com/eagle705/nlp-book/assets/7252598/bc401b1f-41d9-4cd8-b533-39e5c255ed94">

Gradient Accum
- from Gradient accumulation reuses shard memory to scale to out-of-memory datasets. (Source: Cloud Native Machine Learning)
![image](https://github.com/eagle705/nlp-book/assets/7252598/da4ac170-c3f3-49e5-8905-9b06bf8409db)
